curve25519-dalek/src/edwards.rs

```diff
@@ -906,6 +906,23 @@ impl EdwardsPoint {
     ) -> EdwardsPoint {
         crate::backend::vartime_double_base_mul(a, A, b)
     }
+
+    /// Checks whether \\([8a]A + [8b]B = [8]C\\) in variable time.
+    ///
+    /// This can be used to implement [RFC 8032]-compatible Ed25519 signature validation.
+    /// Note that it includes a multiplication by the cofactor.
+    ///
+    /// [RFC 8032]: https://tools.ietf.org/html/rfc8032
+    pub fn vartime_check_double_scalar_mul_basepoint(
```

Comment:
`ed25519-dalek` is now in the same workspace as `curve25519-dalek`, so I can make changes to it in this PR, but I think the next question is how we use this method.

I opened this PR in May 2020. Originally I just returned the scalar mul output directly, but @hdevalence suggested this "check" API instead, where the `EdwardsPoint` version would multiply by the cofactor. I migrated to that, noting that we might want to make the cofactor multiplication configurable.

In October 2020 @hdevalence published his survey of Ed25519 validation criteria. Some time in the intervening 3.5 years, `ed25519-dalek` has gained several separate signature verification methods, that all use this helper function internally:
https://github.com/dalek-cryptography/curve25519-dalek/blob/cc3421a22fa7ee1f557cbe9243b450da53bbe962/ed25519-dalek/src/verifying.rs#L216-L224

These helpers are therefore either checking "ad-hoc" or "strict" equality of R, neither of which multiply by the cofactor. Meanwhile the `ed25519-zebra` crate implements the [ZIP 215](https://zips.z.cash/zip-0215) signature validation rules, which are the "expansive" rules (R is not required to be a canonical encoding, and multiplication by cofactor is required).

So I think we do want some kind of configurability here over the cofactor multiplication. What should this look like? A boolean argument, or two separate APIs?

Note also that the scalar mul optimization implemented in this PR actually checks `[δa]A + [δb]B = [δ]C`, where `δ` is a value invertible mod $\ell$. As @pornin notes in Section 3 of [ePrint 2020/454](https://eprint.iacr.org/2020/454):
> If there is on the curve a non-trivial point `T` of order `h`, then replacing `R` with `R+T` will make
the standard verification equation fail, but the second one will still accept the signature if it
so happens that the value `δ` (obtained from the lattice basis reduction algorithm) turns out to
be a multiple of `h`.

Is there a way we can avoid this by adjusting the lattice basis reduction algorithm to filter out these `δ` values? If not, then we cannot use this optimisation for the "strict" verification methods, and it is debatable whether we should even use them for the "ad-hoc" methods (as doing so would change the ill-defined set of valid signatures - not that there isn't already wide inconsistencies between implementations here already, but this would be a difference between two versions of `curve25519-dalek`, and IDK what the maintainers' policy here is).

Regardless, we definitely should offer a "mul-by-cofactor" version of this in the API, as `curve25519-zebra` (and anyone else using the cofactor check equation) will benefit from it.

Reply:
I have looked a bit at the problem, I think leveraging the optimization while being strictly equivalent to the cofactorless equation is doable, but it is a bit unpleasant.

We have a public key `A`, generator is `B`, signature is `(R, s)`, and during the verification, the challenge `k` is computed as a SHA-512 output, which is then interpreted as an integer. The curve has order `8*L`. Points `A` and `R` are on the curve, but not necessarily in the subgroup of order `L`. The cofactorless verification equation is:

    s*B - k*A = R

First, we should note that while `k` is nominally a 512-bit integer, the implementation in curve25519-dalek represents k as a `Scalar`, which implies reduction modulo `L`. This already deviates from the cofactorless equation in [RFC 8032](https://www.rfc-editor.org/rfc/rfc8032#section-5.1.7), where there is no such reduction. This matters if `A` is not in the subgroup of order `L`; for instance, it may happen that `k` is, as an integer, a multiple of 8, while `k mod L` is an odd integer, in which case the cofactorless equation would report a success, while the dalek implementation would reject it. The reverse is also possible (signature accepted by dalek but rejected by the RFC). All these variants are still within the scope of the signature algorithm, i.e. the discrepancies between verifier behaviours do not allow actual signature forgeries by attackers not knowing the private key. There is some extra discussion in the [Taming the many EdDSAs](https://eprint.iacr.org/2020/1244) paper (page 11). Here I am discussing reproducing the exact behaviour of the current dalek implementation, and therefore I call `k` the reduction of the SHA-512 output modulo `L`.

Given `k`, one can compute `k8 = k mod 8` (the three low bits of `k`). The cofactorless equation is then equivalent to:

    s*B - ((k - k8)/8)*(8*A) - (R + k8*A) = 0

Thus, by replacing `k`, `A` and `R` with, respectively, `(k >> 3)`, `8*A` and `R + (k & 7)*A`, I have a completely equivalent equation (thus with the same behaviour), but I have also guaranteed that the `A` point is in the proper subgroup of order `L`. Thus we can now assume that `A` is in that subgroup. This is important: when multiplying `A` by an integer `x`, we can now reduce `x` modulo `L` without any loss of information.

When we apply Lagrange's algorithm on the lattice basis `((k, 1), (L, 0))`, we get a new basis `((u0, u1), (v0, v1))` for the same lattice. In algorithm 4 in [my paper](https://eprint.iacr.org/2020/454), we stop as soon as the smaller of these two vectors is "small enough", but we can also reuse the stopping condition from algorithm 3, i.e. we can change this:

    if len(N_v) <= t:
        return (v, u)

into:

    if len(N_v) <= t:
        if 2*abs(p) <= N_v:
            return (v, u)

This would, on average, add maybe one or two iterations to the algorithm, i.e. the extra cost on the algorithm would likely be negligible. By using this test, we ensure that not only `v` is truly the smallest non-zero vector in the lattice, but `u` is the second smallest non-zero vector among those which are not colinear to `v` (this kind of assertion breaks down at higher lattice dimensions, but in dimension 2 it works).

Now, Lagrange's algorithm starts here with `u = (k, 1)`, and 1 is odd. Moreover, each step either adds a multiple of `v` to `u`, or a multiple of `u` to `v`. The consequence is that `u1` and `v1` can never both be even; at least one of them is odd. The important point here is that if `v1` is an odd integer, and less than `L` (by construction), then it is invertible modulo `L` (since `L` is prime) but also modulo 8 (since it is odd). Thus, `v1` is invertible modulo `8*L`. If `v1` is invertible modulo `8*L`, which is the whole curve order, then we can multiply the verification equation by `v1` in a reversible way, i.e. without changing the behaviour. We thus get:

    (v1*s mod L)*B - v0*A - v1*R = 0

which is the Antipa _et al_ optimization. Note that the equivalence relies on two properties: that `A` is in the right subgroup (so that we can replace `k*v1` with `v0`), and that `v1` is odd.

**The unpleasantness** is that `v1` might be even. As explained above, if `v1` is even, then `u1` must be odd, hence we can use `(u0, u1)` instead of `(v0, v1)`. However, the smallest non-zero vector in the lattice is `v`, not `u`. Heuristically, `u` is not much bigger than `v`, but there are some degenerate cases. For instance, if `k = (L - 1)/2`, then the output of Lagrange's algorithm is `v = (1, -2)` (very small, but -2 is even), and `u = (2*(L+1)/5, (L-4)/5)` (denominator `u1 = (L-4)/5` is odd, but both `u0` and `u1` are almost as large as `L`).

In the verification algorithm, `k` is an output of SHA-512, and thus attackers would have trouble crafting signatures that leverage the most degenerate cases, and we can heuristically consider that `u` won't be a very large vector, but the lattice reduction algorithm must still performs update on `u` and `v` with their full 254-bit size (including the sign bit); the nice trick of computing them only over 128 bits is no longer applicable. This may conceivably increase the cost, and thus decrease the usefulness of the optimization.

**Summary:** the behaviour of the current implementation (with the cofactorless equation) can be maintained while applying the Antipa _et al_ optimization, provided that the following process is applied:

1. Compute `k` as previously, with a SHA-512 output and with reduction modulo `L` (to maintain backward compatibility).
2. Replace `k`, `A` and `R` with `k >> 3`, `8*A` and `R + (k & 7)*A`, respectively.
3. Compute Lagrange's algorithm over `((k, 1), (L, 0))` (optionally with the extra ending test so that a truly size-reduced basis is obtained, to make both basis vectors as small as possible). Updates to coordinates of `u` and `v` must be maintained over their full size (254 bits).
4. Given the output `((v0, v1), (u0, u1))` of Lagrange's algorithm (with `(v0, v1)` being the smallest non-zero vector in the lattice), use `(v0, v1)` if `v1` is odd; but if `v1` turns out to be even, use `(u0, u1)` instead (in that case, `u1` is odd). Since `u` is not the smallest vector, its coordinates can be larger than `sqrt(1.16*L)`, so the combined Straus algorithm must be able to handle large coefficients (even if these are improbable in practice).

**WARNING:** I wrote all this without actually implementing it. It seems to make sense on paper, but until it is implemented and tested, there's no guarantee I did not make a mistake.

Reply:
Thanks @pornin for looking into this! I think the proposed changes are complex enough that they should be made and tested in a separate PR.

To avoid blocking this PR further, I propose that we rename `EdwardsPoint::vartime_check_double_scalar_mul_basepoint` to something like `EdwardsPoint::vartime_check_double_scalar_mul_basepoint_cofactor`, and then in a subsequent PR we can attempt to expose an `EdwardsPoint::vartime_check_double_scalar_mul_basepoint` that is cofactor-less.